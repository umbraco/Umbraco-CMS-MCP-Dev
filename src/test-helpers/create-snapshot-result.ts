import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
