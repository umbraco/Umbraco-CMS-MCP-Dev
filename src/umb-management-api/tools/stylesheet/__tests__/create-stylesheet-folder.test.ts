import CreateStylesheetFolderTool from "../post/create-stylesheet-folder.js";
import { StylesheetHelper } from "./helpers/stylesheet-helper.js";
import { StylesheetFolderBuilder } from "./helpers/stylesheet-folder-builder.js";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";
import { jest } from "@jest/globals";

const TEST_FOLDER_NAME = "_TestCreateStylesheetFolder";
const EXISTING_FOLDER_NAME = "_ExistingStylesheetFolder";
const TEST_PARENT_FOLDER_NAME = "_TestParentFolder";

describe("create-stylesheet-folder", () => {
  let originalConsoleError: typeof console.error;

  beforeEach(() => {
    originalConsoleError = console.error;
    console.error = jest.fn();
  });

  afterEach(async () => {
    await StylesheetHelper.cleanup(TEST_PARENT_FOLDER_NAME);
    await StylesheetHelper.cleanup(TEST_FOLDER_NAME);
    await StylesheetHelper.cleanup(EXISTING_FOLDER_NAME);
    console.error = originalConsoleError;
  });

  it("should create a stylesheet folder", async () => {
    // Arrange
    const params = {
      name: TEST_FOLDER_NAME
    };

    // Act
    const result = await CreateStylesheetFolderTool().handler(params, { signal: new AbortController().signal });

    // Assert
    const normalizedResult = createSnapshotResult(result);
    expect(normalizedResult).toMatchSnapshot();

    // Verify the created folder exists by path
    const exists = await StylesheetHelper.verifyStylesheetFolder(`/${TEST_FOLDER_NAME}`);
    expect(exists).toBe(true);
  });

  it("should handle existing folder name", async () => {
    // Arrange - First create the folder
    await CreateStylesheetFolderTool().handler({
      name: EXISTING_FOLDER_NAME
    }, { signal: new AbortController().signal });

    // Act - Try to create it again with same name
    const result = await CreateStylesheetFolderTool().handler({
      name: EXISTING_FOLDER_NAME
    }, { signal: new AbortController().signal });

    // Assert
    expect(result).toMatchSnapshot();
  });

  it("should create folder with parent folder", async () => {
    // Arrange - Create parent folder first using builder
    const parentFolderBuilder = new StylesheetFolderBuilder()
      .withName(TEST_PARENT_FOLDER_NAME);
    
    await parentFolderBuilder.create();
    
    const params = {
      name: TEST_FOLDER_NAME,
      parent: {
        path: parentFolderBuilder.getPath()
      }
    };

    // Act
    const result = await CreateStylesheetFolderTool().handler(params, { signal: new AbortController().signal });

    // Assert
    const normalizedResult = createSnapshotResult(result);
    expect(normalizedResult).toMatchSnapshot();
    
    // Verify the nested folder structure
    const exists = await StylesheetHelper.verifyStylesheetFolder(`${parentFolderBuilder.getPath()}/${TEST_FOLDER_NAME}`);
    expect(exists).toBe(true);
  });
});