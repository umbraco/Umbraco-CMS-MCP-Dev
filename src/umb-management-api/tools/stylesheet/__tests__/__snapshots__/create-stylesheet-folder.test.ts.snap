// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-stylesheet-folder should create a stylesheet folder 1`] = `
{
  "content": [
    {
      "text": """",
      "type": "text",
    },
  ],
}
`;

exports[`create-stylesheet-folder should create folder with parent folder 1`] = `
{
  "content": [
    {
      "text": """",
      "type": "text",
    },
  ],
}
`;

exports[`create-stylesheet-folder should handle existing folder name 1`] = `
{
  "content": [
    {
      "text": "Error using create-stylesheet-folder:
{
  "message": "Request failed with status code 400",
  "response": {
    "type": "Error",
    "title": "Folder already exists",
    "status": 400,
    "detail": "The folder already exists",
    "operationStatus": "AlreadyExists"
  }
}",
      "type": "text",
    },
  ],
}
`;
